Add segmented control style

DivineDominion opened the pull request:
Closes https://github.com/sindresorhus/Preferences/issues/2

I struggled a while with different approaches until I threw out `NSTabViewController` out of the window and began managing the toolbar in a custom controller.

Now you can configure the style upon initialization and have it your way. Replicating the transition animation was a bit tricky because the user could click faster than the transition, ending in an invalid view state; I added a `PausableWindow` that swallows user events to prevent this from happening without disabling the toolbar controls.

Enhancements I'd prefer to tackle in separate issues/PRs:

- make the segmented control configurable: `isCentered`/`alignment`, `isSizedUniformly` (all segments have the same with; that's the current state)
- make the toolbar version configurable: `isCentered`/`alignment`, `toolbarStyle` (icon, label, icon and label)
- hide window management of `PreferencesWindowController`: make `PreferencesWindowController` internal, expose `PreferencesController` as a plain Swift object with a simple API
- make preferences controllers configurable after initialization (which also makes it possible again to use `init(fromCoder:)`)

-----

Changes to the PR:

- [X] Extract PausableWindow into its own file and add a comment about what exactly it does and its use-case/purpose
- [X] Use own identifiers:  extension NSToolbarItem.Identifier 
- [X] consider rename crossfadeTransitions to animate
- [X] rename preferencesWindowController.showWindow(withPreference: .advanced)
- [X] add preferencesWindowController.change(to: .preferenceAdvanced) (obsolete)
- [x] change style from tabs to segmented controls on the fly
- [x] left-align toolbar items
- [x] center window around segmented controls

DivineDominion commented:
Something I'd like to have feedback on: using Auto Layout to pin the content views to the window size; because on the one hand, it makes window size management trivial, but on the other you end up with somewhat unexpected constraint animations when the window size change is animated.

DivineDominion commented:
@sindresorhus Did you have a chance to take a look? You want me to adjust something?

sindresorhus commented:
I really appreciate the PR. I’ve just been too busy in real life to look at it yet (wedding and family visits). I’ll hopefully have a chance to look at it the coming week :)

DivineDominion commented:
The latest commit adds a representation for what we may call "overflow menu", which is how AppKit displays segmented controls in toolbars that outgrow the window:

![image](https://user-images.githubusercontent.com/59080/49753945-1ec04f80-fcb5-11e8-8c70-e6a203e70460.png)

In addition, we should also prevent this from happening by keeping a minimum window size, I would say. But nevertheless, if users force the window size down, this is a better fallback than the previous behavior, i.e. displaying an empty menu.

DivineDominion commented:
@sindresorhus Did you have a chance to assess the code so far? :)

DivineDominion commented:
Today's changes make prefs work with Auto Layout properly.

- disable `translatesAutoResizingMAskIntoConstraints` in the preferences views in IB
- add sufficient constraints to define the preferences views' size
- use `NSView.fittingSize` to compute the new window frame during animations, effectively fixing when you save IB files at a size that won't fit the runtime size
- use built-in `NSViewController.transition(from:to:options:completionHandler:)` for animations
- cache and remove constraints when embedding the child view controller, fixing jiggling in transitions

sindresorhus commented:
This is looking really promising. I finally have a chance to look through the code now.

sindresorhus commented:
One thing I noticed is that when using the segmented control style and I change a tab, it should animate from center, meaning, the segmented buttons should not move.

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -1,23 +1,32 @@
 import Cocoa
 import Preferences
 
+extension NSToolbarItem.Identifier {
```
I think we should use our own namespace for the identifiers instead of overloading `NSToolbarItem`. The identifier is for the whole preference pan, not just the toolbar item.

sindresorhus replied:
Having our own namespace also lets us shorten the identifier from `preferenceGeneral` to just `general`.

sindresorhus commented on Sources/Preferences/PreferencesWindowController.swift:
```diff
@@ -18,33 +18,52 @@ public final class PreferencesWindowController: NSWindowController {
 	}
 }
+
+class PausableWindow: NSWindow {
```
Can you move this to its own file?

sindresorhus commented:
I think it could be useful to have a public method to change tabs while the preferences window is open.

Something like:

```swift
preferencesWindowController.change(to: .preferenceAdvanced)
```

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -1,23 +1,32 @@
 
 	func applicationDidFinishLaunching(_ notification: Notification) {
-		preferencesWindowController.showWindow()
+		preferencesWindowController.showWindow(toolbarItemIdentifier: .preferenceAdvanced)
```
I think this would read better as:

```swift
preferencesWindowController.showWindow(withPreference: .advanced)
```

Or something similar.

DivineDominion replied:
Since `showWindow` would do the same as `change(to:)`, what do you think about the latest change to `showPreference`? I hope it brings across that it opens the window if needed, too.

sindresorhus replied:
Yeah, that's good.

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -1,23 +1,32 @@
+		GeneralPreferenceViewController(),
+		AdvancedPreferenceViewController()
+	]
+	lazy var preferencesWindowController: PreferencesWindowController = PreferencesWindowController(preferences: self.preferences, style: .segmentedControl, crossfadeTransitions: true)
```
I would call `crossfadeTransitions` just `animate`.

sindresorhus commented on Example/GeneralPreferenceViewController.swift:
```diff
@@ -3,7 +3,8 @@ import Preferences
 	let toolbarItemTitle = "General"
-	let toolbarItemIcon = NSImage(named: NSImage.preferencesGeneralName)!
+	let toolbarItemIcon: NSImage? = NSImage(named: NSImage.preferencesGeneralName)!
+    let toolbarItemIdentifier: NSToolbarItem.Identifier = .preferenceGeneral
```
Since almost everyone would have a `General` preference tab. I wonder if we should ship with a `PreferenableGeneral` type thing that has all these fields already filled out. My goal is to require the least amount of boilerplate possible to set up a preferences window.

DivineDominion replied:
I think that's a fair point, providing a common set of preferences (General, Advanced, Update, maybe License) -- what do you think about collecting this in a new issue so we can do it in one swoop, solving the localization issue #12  ?

sindresorhus replied:
Sure

DivineDominion replied:
see #13 👍

sindresorhus commented:
@DivineDominion Hey, still interested in finishing this? You are so close :)

DivineDominion commented:
I can have a look at this with one of my next app's iterations. I find it surprisingly hard to animate both the window frame and the content size smoothly, so growing the window symmetrically left and right to make the segmented controls stick could take more time than I can spare in the upcoming week or two.

I'll schedule time for this, though. Thanks for the review so far :) Good to know it's only minor things.

DivineDominion commented:
Started with rebasing my fork on the current master branch.

DivineDominion commented:
- [x] Extract PausableWindow 
- [x] Use own identifiers:  extension NSToolbarItem.Identifier 


The `PreferenceIdentifier` type makes everything nicer except where `NSToolbarItem` is passed to `@IBAction`s or delegate methods. There, I have now added a conversion to the new identifier as:

```swift
PreferenceIdentifier(fromToolbarItemIdentifier: toolbarItem.itemIdentifier)
```

... which works, since they're both backed by Strings, but I think that's rather ugly.

sindresorhus commented:
> The PreferenceIdentifier type makes everything nicer except where NSToolbarItem is passed to @IBActions or delegate methods.

You mean if the user wants to call `.showWindow()` in a @IBActions/delegate method? If so, can't we just add a `.showWindow()` overload that accepts `NSToolbarItem` and do the conversion internally?

sindresorhus commented:
What do you think of renaming `Preferencable` to `Preference` or `PreferencePane`?

> Protocols that describe what something is should read as nouns (e.g. Collection).
> Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).
> \- https://swift.org/documentation/api-design-guidelines/

I feel like a preference pane, *is* something, it doesn't just add a capability, but I might be interpreting it incorrectly.

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -1,14 +1,9 @@
-		return NSToolbarItem.Identifier(rawValue: "preferenceAdvanced")
-	}
+extension PreferenceIdentifier {
+	static let general = PreferenceIdentifier(rawValue: "preferenceGeneral")
```
Can't the raw value `preferenceGeneral` just be `general`? It's already clear from the context what it is.

sindresorhus commented on Example/AppDelegate.swift:
```diff
@@ -1,23 +1,27 @@
 
+extension PreferenceIdentifier {
+	static let general = PreferenceIdentifier(rawValue: "preferenceGeneral")
+	static let advanced = PreferenceIdentifier(rawValue: "preferenceAdvances")
```
```suggestion
	static let advanced = PreferenceIdentifier(rawValue: "preferenceAdvanced")
```

sindresorhus commented:
@DivineDominion Any thoughts on https://github.com/sindresorhus/Preferences/pull/6#discussion_r248296504?

sindresorhus commented on Sources/Preferences/UserInteractionPausableWindow.swift:
```diff
@@ -0,0 +1,31 @@
+        }
+
+        super.sendEvent(event)
+    }
```
Instead of `sendEvent`, couldn't it simply be:

```swift
override func hitTest(_ point: NSPoint) -> NSView? {
	return isUserInteractionEnabled ? super.hitTest(point) : nil
}
```

instead?

DivineDominion replied:
IIRC, keyboard selection (tabbing around, then using the control with space/enter) was an additional problem that led me to use `sendEvent`

sindresorhus replied:
Ah, that makes sense. I didn't think of that interaction.

sindresorhus commented:
We should have a button in the example app to switch between the preference window styles.

sindresorhus commented on Sources/Preferences/Preferenceable.swift:
```diff
@@ -1,6 +1,38 @@
-	var toolbarItemIcon: NSImage { get }
+	var toolbarItemIcon: NSImage? { get }
+	var viewController: NSViewController { get }
+}
```
Currently, if you specify `.segmentedControl` as style, you still have to implement `toolbarItemIcon`, which is weird. I think we should make it truly optional, but adding:

```swift
extension Preferenceable {
	public var toolbarItemIcon: NSImage? {
		return nil
	}
}
```

sindresorhus replied:
Actually, we would have to make it:

```swift
public protocol Preferenceable: AnyObject {
	var preferenceIdentifier: PreferenceIdentifier { get }
	var toolbarItemTitle: String { get }
	var toolbarItemIcon: NSImage { get } // not optional
	var viewController: NSViewController { get }
}

extension Preferenceable {
	public var toolbarItemIcon: NSImage {
		return NSImage(size: .zero)
	}
}
```

That way the user can either specify a non-optional ` let toolbarItemIcon = NSImage(named: NSImage.preferencesGeneralName)!` or nothing at all.

DivineDominion replied:
sounds good 👍

sindresorhus commented on Sources/Preferences/PreferencesWindowController.swift:
```diff
@@ -18,29 +18,24 @@ public final class PreferencesWindowController: NSWindowController {
 	}
 
-	public func showWindow() {
+	public func showWindow(preferenceIdentifier: PreferenceIdentifier? = nil) {
```
What do you think about this? (Or another short word) Would make it more fluent and Swifty.

```suggestion
	public func showWindow(at preferenceIdentifier: PreferenceIdentifier? = nil) {
```

DivineDominion replied:
Would you prefer `showPreference(preferenceIdentifier:)` to skip the parameter name so one can call `showPreference(.general)`? Or rather `show(preference: .general)`?

DivineDominion replied:
I think I like `show(preference:)` most, in fact, now that names and identifiers have changed to be truly unique.

sindresorhus replied:
> I think I like show(preference:) most, in fact, now that names and identifiers have changed to be truly unique.

👍 

But are we sure `preference` is clear enough? "Preference" could really mean **one setting*, as in **preferences** are many settings (user preferences), and **a preference** is one setting. So maybe we should go with "preferencePane" for everything? Or maybe you have a better naming?

DivineDominion replied:
That reminds me of the initial naming struggle :)

Looked it up this time. Leaning on to the [Apple HIG](https://developer.apple.com/design/human-interface-guidelines/macos/app-architecture/preferences/), I think

- `Preference` is 1 setting (e.g. corresponds to a single UserDefaults key)
- `Preferences` are for the app's whole collection of settings, as per the HIG:
    > A **preferences** window is a modeless dialog [...]
- `PreferencePane` for the view controller pages
    >  Update the window's title to reflect the currently visible **preference pane.**

So `PreferencePane` works for me.

sindresorhus commented:
I noticed that when using the `.tabs` style, the tabs are centered, but they should be left aligned.

DivineDominion commented:
> > The PreferenceIdentifier type makes everything nicer except where NSToolbarItem is passed to @IBActions or delegate methods.
> 
> You mean if the user wants to call `.showWindow()` in a @IBActions/delegate method? If so, can't we just add a `.showWindow()` overload that accepts `NSToolbarItem` and do the conversion internally?

No, just the internal stuff like https://github.com/sindresorhus/Preferences/pull/6/commits/0cdad7cd52f4173ea6f313e87760a20798ac6e55#diff-f1852e181127fd79b06bc13906608fbdR158 -- I mean, it works, but I don't like the cross-conversion in both ways.

DivineDominion commented:
I added a list of open tasks to the original PR message at the top to keep track of things.

sindresorhus commented:
> We should have a button in the example app to switch between the preference window styles.

I actually didn't mean it had to be dynamically changeable. That doesn't really make sense outside the example app, so we don't really have to complicate the code for that. I was thinking just restarting the app when the button is pressed. Should be almost instantaneous anyway.

DivineDominion commented:
> I actually didn't mean it had to be dynamically changeable. That doesn't really make sense outside the example app, so we don't really have to complicate the code for that. I was thinking just restarting the app when the button is pressed. Should be almost instantaneous anyway.

Gotcha. Changed it.

Also fixed the remaining issues. Centering the window is jiggling the segmented control around a bit, probably due to rounding errors. I tried `floor` and `ceil` in `PreferencesTabViewController.setWindowFrame(for:animated:)`, but that didn't help either.

sindresorhus commented on Sources/Preferences/ToolbarItemStyleViewController.swift:
```diff
@@ -7,7 +7,7 @@ final class ToolbarItemStyleViewController: NSObject, PreferencesStyleController
     weak var delegate: PreferencesStyleControllerDelegate?
 
-    init(preferences: [PreferencePane], toolbar: NSToolbar, centerToolbarItems: Bool = true) {
+    init(preferences: [PreferencePane], toolbar: NSToolbar, centerToolbarItems: Bool) {
```
I think `centerToolbarItems` should default to `false` here. No point in forcing users to specify that value when most will want the default of not centering.

DivineDominion replied:
Oh, did you notice it's not part of the public API? Since this is only internal, I got rid of the default to highlight that the parameter exists at all. More like a humble reminder for future changes than something users will see. Still want me to change this?

sindresorhus replied:
I missed that. Ignore my comment.

sindresorhus commented:
From the HIG:

> Restore the last viewed preference pane. If the user switches preference panes, your app should remember this change and show the same pane immediately the next time the user opens your preferences window.

We don't currently do that. I think we should do it when users call `.show()` without an argument, and mention in the docs that that should be the preferred way to show the preferences window, and refer to the HIG docs. This would be another benefit of having https://github.com/sindresorhus/Preferences/issues/14 and we could enforce it there.

DivineDominion commented:
@sindresorhus  Will happily add that, but again favor doing that in another PR to keep this focused on the segmented controls. Would that work with you?

sindresorhus commented:
Yup, agreed. This PR is already very large. Moved it to a new issue: https://github.com/sindresorhus/Preferences/issues/16

sindresorhus commented:
Can you update the readme to reflect the changes?

sindresorhus commented:
> Also fixed the remaining issues. Centering the window is jiggling the segmented control around a bit, probably due to rounding errors. I tried floor and ceil in PreferencesTabViewController.setWindowFrame(for:animated:), but that didn't help either.

I don't really know a solution to that either. If you cannot find a solution, I'm fine with just deferring it to later and move it to a new issue.

DivineDominion commented:
I think I caught everything in the README. I also added screenshots of the two styles, but oddly the README preview on my branch doesn't work. I hope it's just due to it being a fork and relative paths not working correctly, but can you double-check this when you merge?

sindresorhus commented:
Didn’t we decide on `.show(preferencePane: .general)`? The word “preferences” in `.showPreferences` feels moot as it’s already part of the controller name and the parameter. We also need to mention that we recommend `.show()` without an argument.

sindresorhus commented:
Otherwise, the changes look great! 👌

DivineDominion commented:
You're right, this totally went under my radar. I added a docstring for now; when #16 is implemented, the recommendation can be expanded and will make more sense IMHO

sindresorhus commented:
This is absolutely amazing. I'm very happy about the end result. Thank you so much for working so hard on this, @DivineDominion. I really appreciate it. 🙌

Should we do a new release with what we have here now or do you intend to work on more things in the near future?

Also, would you be interested in joining the project as a maintainer? No commitment required. Just want to highlight your contribution in the readme.

sindresorhus commented on readme.md:
```diff
@@ -99,37 +112,70 @@ import Preferences
+         style: PreferencesStyle = .toolbarItems, 
+         animated: Bool = true)
+    func show(preferencePane: PreferencePaneIdentifier? = nil)
+    func hideWindow()
```
I just noticed we have this. It's a bit inconsistent with `.show()` and `.hideWindow()`. Maybe we should add a `.hide()` method too?

DivineDominion replied:
Yeah, I stumbled upon that, too. AppKit uses showWindow/close by default, and we now have our show/hideWindow in addition to that. Why not just add the `show(preferencePane:)` variant in addition to `showWindow(_:)`, and stick to `close`?

sindresorhus replied:
So just remove `hideWindow()` and document `close()`? If so, I agree. I think it's also worth documenting `close()` just for convenience of the user.

DivineDominion replied:
Ok, see #21

sindresorhus commented:
I forgot to double-check the images before merge, but they are definitely corrupt. Would you mind submitting them again?

DivineDominion commented:
I was transitioning all my macOS apps to this framework, so I have sufficient skin in the game to help you with this. If you can use a helping hand, I'd be happy to contribute!

I don't have a strong opinion about releases. In fact, I prefer more frequent releases over big releases that take a long time. Adding this is kind of a big thing, compared to the original code base, so why not! Then fix the open issues and localization problems.

sindresorhus commented:
> I was transitioning all my macOS apps to this framework, so I have sufficient skin in the game to help you with this. If you can use a helping hand, I'd be happy to contribute!

When you have done so, I would be happy to list your apps in the readme.

sindresorhus commented:
I also replied to https://github.com/sindresorhus/Preferences/pull/6#discussion_r272485257, in case it gets lost.

DivineDominion commented:
Currently using this for:

- https://zettelkasten.de/the-archive/
- https://wordcounterapp.com/

Am transitioning this one when we're done with the next release:
- https://christiantietze.de/move-work-break/

sindresorhus commented:
> Cool! Am currently using this for:

Can you do a PR to add them to the readme? Maybe a new section above the FAQ section.
